src/compiler/scala/tools/nsc/transform/patmat/MatchTranslation.scala

```diff
@@ -409,22 +409,22 @@ trait MatchTranslation {
       protected def expectedLength      = elementArity
       protected def lastIndexingBinder  = nonStarArity - 1
 
-      private def productElemsToN(binder: Symbol, n: Int): List[Tree] = 1 to n map tupleSel(binder) toList
-      private def genTake(binder: Symbol, n: Int): List[Tree]         = (0 until n).toList map (codegen index seqTree(binder))
-      private def genDrop(binder: Symbol, n: Int): List[Tree]         = codegen.drop(seqTree(binder))(expectedLength) :: Nil
+      private def productElemsToN(binder: Symbol, n: Int): List[Tree] = if (n == 0) Nil else 1 to n map tupleSel(binder) toList
+      private def genTake(binder: Symbol, n: Int): List[Tree]         = if (n == 0) Nil else (0 until n).toList map (codegen index seqTree(binder, forceImmutable = false))
```

Comment:
Seems we could avoid creating both a Range + a List + another List with `List.tabulate(n)(codegen index seqTree(binder, forceImmutable = false))` ?

Reply:
Thanks, pushed that suggestion.